// vite-plugin-checker/lib/checkers/vls/initParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultVLSConfig = exports.getInitParams = void 0;
function getInitParams(workspaceUri) {
    const defaultVLSConfig = getDefaultVLSConfig();
    defaultVLSConfig.vetur.validation = {
        template: true,
        style: true,
        script: true,
        interpolation: true,
        templateProps: true,
    };
    defaultVLSConfig.vetur.experimental = {
        templateInterpolationService: true,
    };
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    const init = {
        rootPath: workspaceUri.fsPath,
        rootUri: workspaceUri.toString(),
        processId: process.pid,
        capabilities: {},
        initializationOptions: {
            config: defaultVLSConfig,
        },
    };
    return init;
}
exports.getInitParams = getInitParams;
function getDefaultVLSConfig() {
    return {
        vetur: {
            ignoreProjectWarning: false,
            useWorkspaceDependencies: false,
            validation: {
                template: true,
                templateProps: true,
                interpolation: true,
                style: true,
                script: true,
            },
            completion: {
                autoImport: false,
                tagCasing: 'initial',
                scaffoldSnippetSources: {
                    workspace: '💼',
                    user: '🗒️',
                    vetur: '✌',
                },
            },
            grammar: {
                customBlocks: {},
            },
            format: {
                enable: true,
                options: {
                    tabSize: 2,
                    useTabs: false,
                },
                defaultFormatter: {},
                defaultFormatterOptions: {},
                scriptInitialIndent: false,
                styleInitialIndent: false,
            },
            languageFeatures: {
                codeActions: true,
                updateImportOnFileMove: true,
                semanticTokens: true,
            },
            trace: {
                server: 'off',
            },
            dev: {
                vlsPath: '',
                vlsPort: -1,
                logLevel: 'INFO',
            },
            experimental: {
                templateInterpolationService: false,
            },
        },
        css: {},
        html: {
            suggest: {},
        },
        javascript: {
            format: {},
        },
        typescript: {
            tsdk: null,
            format: {},
        },
        emmet: {},
        stylusSupremacy: {},
    };
}
exports.getDefaultVLSConfig = getDefaultVLSConfig;
